File: browse-cloudflare-access-pages/rest/php/cf-access.php
<?php
// Two approaches for accessing Cloudflare Access-protected pages via the REST API:
//   1. Saved profile — reuse a browser session captured after logging in through CF Access.
//   2. Service Token — inject CF-Access headers via setExtraHTTPHeaders for machine-to-machine access.
//
// Uses PHP's built-in curl — no Composer packages needed.
//
// Run: php cf-access.php

$token = 'YOUR_API_TOKEN_HERE';

// Approach 1: reuse a saved authenticated profile.
$ch = curl_init("https://production-sfo.browserless.io/screenshot?token={$token}&profile=cf-access-profile");
curl_setopt_array($ch, [
    CURLOPT_POST           => true,
    CURLOPT_HTTPHEADER     => ['Content-Type: application/json'],
    CURLOPT_POSTFIELDS     => json_encode(['url' => 'https://internal.example.com/dashboard']),
    CURLOPT_RETURNTRANSFER => true,
]);
$png = curl_exec($ch);
curl_close($ch);
file_put_contents('dashboard.png', $png);
echo "Saved dashboard.png\n";

// Approach 2: inject Service Token headers via setExtraHTTPHeaders.
$ch = curl_init("https://production-sfo.browserless.io/content?token={$token}");
curl_setopt_array($ch, [
    CURLOPT_POST           => true,
    CURLOPT_HTTPHEADER     => ['Content-Type: application/json'],
    CURLOPT_POSTFIELDS     => json_encode([
        'url'                  => 'https://internal.example.com/dashboard',
        'setExtraHTTPHeaders'  => [
            'CF-Access-Client-Id'     => 'YOUR_CF_CLIENT_ID.access',
            'CF-Access-Client-Secret' => 'YOUR_CF_CLIENT_SECRET',
        ],
    ]),
    CURLOPT_RETURNTRANSFER => true,
]);
$content = curl_exec($ch);
curl_close($ch);
echo substr($content, 0, 200) . "\n";

File: browse-cloudflare-access-pages/rest/python/cf-access.py
# Two approaches for accessing Cloudflare Access-protected pages with Playwright:
#   1. Service Token — inject CF-Access headers via extra_http_headers on the context.
#   2. Saved profile — reuse a browser session captured after logging in through CF Access.
#
# Install: pip install playwright
# Run:     python cf-access.py

from playwright.sync_api import sync_playwright

TOKEN = 'YOUR_API_TOKEN_HERE'

# Approach 1: inject Service Token headers.
with sync_playwright() as playwright:
    WS_ENDPOINT = f'wss://production-sfo.browserless.io/chromium/playwright?token={TOKEN}'
    browser = playwright.chromium.connect(WS_ENDPOINT)
    try:
        context = browser.new_context(
            extra_http_headers={
                'CF-Access-Client-Id': 'YOUR_CF_CLIENT_ID.access',
                'CF-Access-Client-Secret': 'YOUR_CF_CLIENT_SECRET',
            }
        )
        page = context.new_page()
        page.goto('https://internal.example.com/dashboard')
        print('Title:', page.title())
    finally:
        # Always close to release the session even on error.
        browser.close()

# Approach 2: reuse a saved authenticated profile.
# connect_over_cdp preserves the saved profile's context and cookies.
with sync_playwright() as playwright:
    WS_ENDPOINT = f'wss://production-sfo.browserless.io?token={TOKEN}&profile=cf-access-profile'
    browser = playwright.chromium.connect_over_cdp(WS_ENDPOINT)
    try:
        context = browser.contexts[0]
        page = context.pages[0]
        page.goto('https://internal.example.com/dashboard')
        print('Title:', page.title())
    finally:
        # Always close to release the session even on error.
        browser.close()
